backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java

```diff
@@ -838,9 +842,36 @@ public FutureExperimentRunDAO(
                                     futureDatasets,
                                     (stream, datasets) ->
                                         stream.map(
-                                            builder ->
-                                                builder.addAllDatasets(
-                                                    datasets.get(builder.getId()))),
+                                            builder -> {
+                                              List<Artifact> datasetList =
+                                                  datasets.get(builder.getId());
+                                              if (datasetList != null && !datasetList.isEmpty()) {
+                                                if (config.populateConnectionsBasedOnPrivileges) {
+                                                  try {
+                                                    return filterAndGetPrivilegedDatasetsOnly(
+                                                            datasetList, true)
+                                                        .thenCompose(
+                                                            artifacts -> {
+                                                              builder
+                                                                  .clearDatasets()
+                                                                  .addAllDatasets(artifacts);
+                                                              return InternalFuture
+                                                                  .completedInternalFuture(builder);
+                                                            },
+                                                            executor)
+                                                        .get();
```

Comment:
Why are you resolving the future immediately? This is extremely bad practice. You're just turning parallel code and making it blocking, which is the whole problem we're trying to solve =)

Reply:
Resolved this on the based PR so this logic has been removed from here